airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py

```diff
@@ -86,19 +86,34 @@ def accumulate(self, record: AirbyteRecordMessage) -> None:
         """Uses the input record to add to the inferred schemas maintained by this object"""
         self.stream_to_builder[record.stream].add_object(record.data)
 
+    def _null_type_in_any_of(self, node: InferredSchema) -> bool:
+        if "anyOf" in node:
+            return {"type": _NULL_TYPE} in node["anyOf"]
+        else:
+            return False
+
+    def _remove_type_from_any_of(self, node: InferredSchema) -> InferredSchema:
+        if "anyOf" in node:
+            node.pop("type", None)
+        return node
+
     def _clean(self, node: InferredSchema) -> InferredSchema:
         """
         Recursively cleans up a produced schema:
         - remove anyOf if one of them is just a null value
         - remove properties of type "null"
         """
+
         if isinstance(node, dict):
             if "anyOf" in node:
-                if len(node["anyOf"]) == 2 and {"type": _NULL_TYPE} in node["anyOf"]:
+                if len(node["anyOf"]) == 2 and self._null_type_in_any_of(node):
```

Comment:
Somewhat tangential question: why is this checking that the length of `node["anyOf"]` is `2`? What if there are three or more values in the `anyOf`? It seems like we'd want to handle that case here as well, right?

Reply:
@lmossman do we allow for multiple types? My understanding is that len == 2 means there is a type + null

Reply:
In the [OC issue](https://github.com/airbytehq/oncall/issues/5691) that prompted this PR, the input data is
```
{
    "attributes": [
        {
            "title": "Usual Clothing Size",
            "type": "multi-value",
            "value": [
                "XL"
            ]
        },
        {
            "title": "Where do you live?",
            "type": "location",
            "value": {
                "countryCode": "GB",
                "countryName": "United Kingdom"
            }
        }
    ]
}
```
which is resulting in this inferred schema for the `attributes` property:
```
{
    "anyOf": [
        {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        {
            "type": "object",
            "properties": {
                "countryCode": {
                    "type": "string"
                },
                "countryName": {
                    "type": "string"
                }
            }
        }
    ]
}
```
so that is an example where we have an `anyOf` with one of the values _not_ being `type: null`. So I could imagine a case where that `value` field had yet another value type in the data (e.g. a number) that would result in the inferred `anyOf` containing 3 different values

Reply:
@lmossman is that even supported by destinations? My feeling is the connector's output should be consistent - in this case, emitting only arrays of strings, and wrapping the singleton strings as necessary.

CC @edgao do destinations support fields with multiple types (eg `Union[str, List[str]]`)?

Reply:
tl;dr I think this is fine, with some caveats

DV2 destinations current behavior:
* `{oneOf: [...]}` is translated to the destination's JSON type (jsonb/super/variant/etc). So this will work reasonably well, though the UX isn't the greatest
* `{anyOf: [...]}` will be treated as an "unrecognized" type, which also just defaults to the destination's JSON type. So it'll behave the same as oneOf, just for different reasons
* `{type: [something_not_null, something_else_not_null, ...]}` is handled using legacy logic ported from normalization (it picks the "widest" type amongst the options)
    * this logic is kind of dumb, e.g. `type: [boolean, timestamp]` becomes a timestamp
    * but even if it picks a dumb type, we'll just null out offending values and add an entry to `airbyte_meta`, so this doesn't strictly block syncs

in the distant future we'll probably (a) make all of those behave identically, and (b) pick the smallest type that can hold all of the unioned types. Not happening anytime soon though.

statically-typed file destinations (i.e. gcs/s3 in avro/parquet mode) already have handling for `oneOf` (... and we want to rewrite these destinations anyway).

references:
* DV2 JsonSchema parser https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/typing_deduping/AirbyteType.kt#L25
* file destinations JsonSchema parser (we want to kill this and switch all destinations to use the DV2 schema parser, just fyi) https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/avro/JsonToAvroSchemaConverter.kt